File: scrapers/betano.js
/**
 * Betano — odds 1x2 ("Resultado Final") via API JSON interna.
 *
 * Endpoints:
 *   /api/sports/FOOT/hot/trending/leagues             → lista de ligas em destaque
 *   /api/sports/FOOT/hot/trending/leagues/{id}/events → eventos da liga com `markets`
 *
 * Cada evento traz `markets[]`; pegamos o de `type === 'MR12'` (com SuperOdds turbinada)
 * preferencialmente, senão `MRES` (Resultado Final padrão). 3 selections por mercado.
 */
import { makeEntry, isFuture, sleep } from './base.js';

const NAME = 'Betano';
const HOST = 'https://www.betano.bet.br';
const SESSION_URL = `${HOST}/sport/futebol/`;
const PARAMS = 'req=s,stnf,c,mb';

export async function scrapeBetano(browser) {
  const results = [];
  const page = await browser.newPage();
  await page.setUserAgent('Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36');
  await page.setExtraHTTPHeaders({
    'Accept-Language': 'pt-BR,pt;q=0.9,en;q=0.8',
    'Accept': 'application/json, text/plain, */*',
  });

  try {
    console.log(`[${NAME}] Sessão em ${SESSION_URL}`);
    // domcontentloaded é ~30s mais rápido que networkidle2; sleep curto cobre cookies.
    try {
      await page.goto(SESSION_URL, { waitUntil: 'domcontentloaded', timeout: 25000 });
    } catch {
      console.log(`[${NAME}] goto timeout, prosseguindo`);
    }
    await sleep(2500);

    // 1) Trending leagues — 2 tentativas (era 4)
    let leagues = [];
    for (let attempt = 1; attempt <= 2; attempt++) {
      const result = await page.evaluate(async (host, params) => {
        try {
          const r = await fetch(`${host}/api/sports/FOOT/hot/trending/leagues?${params}`, { credentials: 'include' });
          if (r.status !== 200) return { status: r.status, leagues: [] };
          const j = await r.json();
          return { status: 200, leagues: j.data?.leagues || j.data || [] };
        } catch { return { status: 0, leagues: [] }; }
      }, HOST, PARAMS);
      if (result.status === 200 && Array.isArray(result.leagues) && result.leagues.length) {
        leagues = result.leagues;
        break;
      }
      console.log(`[${NAME}] tentativa ${attempt}: status=${result.status}, aguardando...`);
      await sleep(3000);
    }

    if (!Array.isArray(leagues) || !leagues.length) {
      console.warn(`[${NAME}] sem trending leagues`);
      return results;
    }

    // 2) Eventos por liga — todas em PARALELO (era sequencial = 5 min)
    const leagueResults = await Promise.all(leagues.map(league =>
      page.evaluate(async (host, params, id) => {
        try {
          const r = await fetch(`${host}/api/sports/FOOT/hot/trending/leagues/${id}/events?${params}`, { credentials: 'include' });
          const j = await r.json();
          return j.data?.events || [];
        } catch { return []; }
      }, HOST, PARAMS, league.id).then(events => ({ league, events }))
    ));

    for (const { league, events } of leagueResults) {
      let added = 0;
      for (const ev of events) {
        const dt = ev.startTime ? new Date(ev.startTime) : null;
        if (dt && !isFuture(dt)) continue;

        // Prioriza MR12 (com SuperOdd aplicada), senão MRES (Resultado Final padrão)
        const market = (ev.markets || []).find(m => m.type === 'MR12')
                    || (ev.markets || []).find(m => m.type === 'MRES');
        if (!market || !Array.isArray(market.selections) || market.selections.length < 2) continue;

        const eventRaw = (ev.name || ev.shortName || '').replace(/\s*-\s*/, ' x ').trim();
        const leagueName = ev.leagueName || ev.leagueDescription || league.name || '';

        for (const sel of market.selections) {
          const odd = sel.price;
          if (!(odd > 1.01)) continue;
          results.push(makeEntry({
            bookmaker: NAME,
            eventRaw,
            league: leagueName,
            sport: 'football',
            eventDatetime: dt,
            market: '1x2',
            selection: sel.fullName || sel.name,
            oddBoosted: odd,
            oddBase: null,
          }));
          added++;
        }
      }
      console.log(`[${NAME}] liga "${league.name}" (${league.id}): ${events.length} events, ${added} entries 1x2`);
    }
  } catch (e) {
    console.error(`[${NAME}] Erro: ${e.message}`);
  } finally {
    try { await page.close(); } catch {}
  }

  console.log(`[${NAME}] ${results.length} odds coletadas`);
  return results;
}

File: scrapers/betano.py
"""
Betano — SuperOdd identificada pelo badge "SO" nas listagens de jogos.
Varre futebol BR + NBA e coleta apenas eventos com o badge ativo.
"""
import logging
import re
from datetime import datetime
from zoneinfo import ZoneInfo
from playwright.async_api import BrowserContext
from .base import BaseScraper

logger = logging.getLogger(__name__)
BRT = ZoneInfo("America/Sao_Paulo")

URLS = [
    ("football", "https://betano.bet.br/sport/futebol/"),
    ("basketball", "https://betano.bet.br/sport/basquetebol/nba/"),
]

# Seletores — inspecione o DOM se algo mudar
# O badge "SO" fica dentro do botão de odd
SEL_WAIT      = "[class*='selections'], [class*='event-row']"
SEL_EVENT     = "[class*='event-row'], [class*='EventRow']"
SEL_BADGE     = "[class*='so-label'], [class*='SoLabel'], [data-qa='so-label']"
SEL_TEAMS     = "[class*='event__participants'], [class*='EventParticipants'], [class*='teams']"
SEL_ODD_BTN   = "[class*='selections__selection'], [class*='Selection']"
SEL_ODD_VAL   = "[class*='odd__value'], [class*='OddValue'], [class*='price']"
SEL_ODD_BASE  = "[class*='original-odd'], [class*='OriginalOdd'], s, [class*='strikethrough']"
SEL_MARKET    = "[class*='market-name'], [class*='MarketName']"
SEL_DATE      = "[class*='event__start-time'], [class*='StartTime'], time"
SEL_LEAGUE    = "[class*='league-name'], [class*='LeagueName'], [class*='competition']"


class BetanoScraper(BaseScraper):
    name = "Betano"

    async def scrape(self, context: BrowserContext) -> list[dict]:
        results: list[dict] = []

        for sport, url in URLS:
            try:
                page = await context.new_page()
                logger.info("[Betano] Abrindo %s", url)
                await page.goto(url, wait_until="domcontentloaded", timeout=30_000)
                await page.wait_for_timeout(4_000)

                # Scroll para carregar lazy-loaded events
                await _scroll_page(page)

                events = await self.wait_and_get_all(page, SEL_EVENT)
                logger.info("[Betano] %d eventos em %s", len(events), sport)

                for ev in events:
                    try:
                        # Só processa se tiver badge SO
                        badge = await ev.query_selector(SEL_BADGE)
                        # Fallback: buscar por texto "SO" em qualquer span dentro do evento
                        if not badge:
                            badges = await ev.query_selector_all("span, [class*='label']")
                            for b in badges:
                                txt = (await b.inner_text()).strip().upper()
                                if txt in ("SO", "SUPER ODD", "SUPERODDS"):
                                    badge = b
                                    break
                        if not badge:
                            continue

                        teams_el = await ev.query_selector(SEL_TEAMS)
                        teams_raw = (await teams_el.inner_text()).strip() if teams_el else ""

                        date_el = await ev.query_selector(SEL_DATE)
                        date_raw = (await date_el.get_attribute("datetime") or
                                    await date_el.inner_text()).strip() if date_el else ""
                        event_dt = _parse_date(date_raw)

                        if not self.is_future(event_dt):
                            continue

                        league_el = await ev.query_selector(SEL_LEAGUE)
                        league = (await league_el.inner_text()).strip() if league_el else ""

                        # Coleta todos os botões de odd com badge SO
                        odd_btns = await ev.query_selector_all(SEL_ODD_BTN)
                        for btn in odd_btns:
                            b = await btn.query_selector(SEL_BADGE)
                            if not b:
                                spans = await btn.query_selector_all("span, [class*='label']")
                                for s in spans:
                                    t = (await s.inner_text()).strip().upper()
                                    if t in ("SO", "SUPER ODD"):
                                        b = s
                                        break
                            if not b:
                                continue

                            odd_el = await btn.query_selector(SEL_ODD_VAL)
                            odd_raw = (await odd_el.inner_text()).strip() if odd_el else ""
                            odd_boosted = self.parse_odd(odd_raw)
                            if not odd_boosted:
                                continue

                            base_el = await btn.query_selector(SEL_ODD_BASE)
                            odd_base = None
                            if base_el:
                                odd_base = self.parse_odd((await base_el.inner_text()).strip())

                            market_el = await ev.query_selector(SEL_MARKET)
                            market = (await market_el.inner_text()).strip() if market_el else "1X2"

                            sel_name = await btn.inner_text()
                            sel_name = re.sub(r"\d+[\.,]\d+", "", sel_name).strip()

                            results.append(self.make_entry(
                                event_raw=teams_raw,
                                event_normalized=self.normalize_name(teams_raw),
                                league=league,
                                sport=sport,
                                event_datetime=event_dt,
                                market=market,
                                selection=sel_name or teams_raw,
                                odd_boosted=odd_boosted,
                                odd_base=odd_base,
                            ))

                    except Exception as e:
                        logger.debug("[Betano] Erro em evento: %s", e)

            except Exception as e:
                logger.error("[Betano] Falha em %s: %s", url, e)
            finally:
                await page.close()

        logger.info("[Betano] %d odds coletadas no total", len(results))
        return results


async def _scroll_page(page):
    for _ in range(5):
        await page.evaluate("window.scrollBy(0, window.innerHeight)")
        await page.wait_for_timeout(800)


def _parse_date(raw: str) -> datetime | None:
    if not raw:
        return None
    # ISO format: "2026-05-05T18:00:00"
    try:
        dt = datetime.fromisoformat(raw.replace("Z", "+00:00"))
        return dt.astimezone(BRT)
    except ValueError:
        pass
    # Betano pode usar "05/05 18:00"
    now = datetime.now(tz=BRT)
    m = re.search(r"(\d{2}/\d{2})\s+(\d{2}:\d{2})", raw)
    if m:
        try:
            dt = datetime.strptime(f"{m.group(1)}/{now.year} {m.group(2)}", "%d/%m/%Y %H:%M")
            return dt.replace(tzinfo=BRT)
        except ValueError:
            pass
    return None
